keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,464 @@
+      
+    ```yaml
+    properties:
+      someArray:
+        type: array
+        x-kubernetes-list-type: map
+        x-kubernetes-list-map-keys: ["name"]
+        x-kubernetes-addition-only: true # allowed with `x-kubernetes-list-type: map`
+        items:
+          properties:
+            name:
+              type: string
+              x-kubernetes-immutable: true # enforced
+            x:
+              type: string
+            y:
+              type: integer
+    ```
+    and similarly for `x-kubernetes-deletion-only: true`.
+      
+  - For arrays of list-type `set` get:
+  
+    ```yaml
+    properties:
+      someSet:
+        type: array
+        x-kubernetes-list-type: set
+        x-kubernetes-addition-only: true # allowed with `x-kubernetes-list-type: set`
+        items:
+          type: object
+          properties:
+            x:
+              type: string
+            y:
+              type: integer
+    ```
+    and similarly for `x-kubernetes-deletion-only: true`.
+    
+For list type `atomic` we disallow `x-kubernetes-immutable-keys`.
+
+Note: it is planned to add `x-kubernetes-map-type` as the equivalent vendor extensions for map to distinguish between classical and atomic maps with the values `map` and `atomic`. The `map` case is the same as the one without any specified map type, and for atomic maps we disallow `x-kubernetes-immutable-keys` again.
+    
+### Publishing
+
+The `x-kubernetes-immutable` and `x-kubernetes-immutable-keys` vendor extensions are published via `/openapi/v2` as is.
+
+### Suggested marker syntax
+
+In analogy to `+required`, `+optional` we propose to add a marker to kubebuilder's
+controller-gen named `+immutable`, which covers the whole subtree of the object 
+(compare field-selection section for extension).
+
+```
+// The name can not be changed after creation.
+// +immutable
+Name string
+
+// The list of containers can not change AT ALL after creation.
+// No single field in existing containers can be changed, added or deleted,
+// no new containers can be added, no existing container can be removed.
+// +immutable
+Containers []Containers
+```
+
+The `x-kubernetes-immutable-keys: true` vendor extension is expressed via `// +immutable=keys` in Golang types:
+
+### Future outline sketch: native resources
+
+For native resources we add a pre-immutability-check normalization step for objects 
+decoded from JSON which have normalizations defined:
+
+1. versioned JSON blob comes in a request
+2. unmarshalled into versioned Golang struct
+3. defaulting
+4. conversion to internal
+5. if immutability is enabled for the resource:
+   1. marshalling into JSON in-memory
+   2. normalization creating a copy with shared data-structures
+   3. strict immutability check against the old object, coming from proto assuming it is normalized.
+
+### Future outline: protobuf
+
+Protobuf encoding unifies unset, empty and null values for slices and maps. Those three cases 
```

Comment:
We should never use the difference between unset, empty, null to drive a semantic behavior difference. We have in the past, and it's been a mistake every time.

In addition to users not expecting that, it breaks apply, because in apply, we need to interpret unset/empty/null as meaning "I have no opinion".